pkg/idtools/idtools.go

```diff
@@ -0,0 +1,184 @@
+		}
+		uid = xUID
+	}
+	if gidMap != nil {
+		xGID, err := ToHost(0, gidMap)
+		if err != nil {
+			return -1, -1, err
+		}
+		gid = xGID
+	}
+	return uid, gid, nil
+}
+
+// ToContainer takes an id mapping, and uses it to translate a
+// host ID to the remapped ID. If no map is provided, then the translation
+// assumes a 1-to-1 mapping and returns the passed in id
+func ToContainer(hostID int, idMap []IDMap) (int, error) {
+	if idMap == nil {
+		return hostID, nil
+	}
+	for _, m := range idMap {
+		if (hostID >= m.HostID) && (hostID <= (m.HostID + m.Size - 1)) {
+			contID := m.ContainerID + (hostID - m.HostID)
+			return contID, nil
+		}
+	}
+	return -1, fmt.Errorf("Host ID %d cannot be mapped to a container ID", hostID)
+}
+
+// ToHost takes an id mapping and a remapped ID, and translates the
+// ID to the mapped host ID. If no map is provided, then the translation
+// assumes a 1-to-1 mapping and returns the passed in id #
+func ToHost(contID int, idMap []IDMap) (int, error) {
+	if idMap == nil {
+		return contID, nil
+	}
+	for _, m := range idMap {
+		if (contID >= m.ContainerID) && (contID <= (m.ContainerID + m.Size - 1)) {
+			hostID := m.HostID + (contID - m.ContainerID)
+			return hostID, nil
+		}
+	}
+	return -1, fmt.Errorf("Container ID %d cannot be mapped to a host ID", contID)
+}
+
+// CreateIDMappings takes a requested user and group name and
+// using the data from /etc/sub{uid,gid} ranges, creates the
+// proper uid and gid remapping ranges for that user/group pair
+func CreateIDMappings(username, groupname string) ([]IDMap, []IDMap, error) {
+	uidStart, uidLength, err := parseSubuid(username)
+	if err != nil {
+		return nil, nil, err
+	}
+	gidStart, gidLength, err := parseSubgid(groupname)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return createIDMap(uidStart, uidLength), createIDMap(gidStart, gidLength), nil
+}
+
+func createIDMap(start, length int) []IDMap {
+	idMap := []IDMap{}
+
+	// create a single continguous map of 65536 length (even if
+	// the sub{uid,gid} range is larger) starting at root (0)
+	idMap = append(idMap, IDMap{
+		ContainerID: 0,
+		HostID:      start,
+		Size:        int(math.Min(65536, float64(length))),
+	})
+	return idMap
+}
+
+func parseSubuid(username string) (int, int, error) {
+	return parseSubidFile(subuidFileName, username)
+}
+
+func parseSubgid(username string) (int, int, error) {
+	return parseSubidFile(subgidFileName, username)
+}
+
+func parseSubidFile(path, username string) (int, int, error) {
+	subidFile, err := os.Open(path)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer subidFile.Close()
+
+	s := bufio.NewScanner(subidFile)
+	for s.Scan() {
+		if err := s.Err(); err != nil {
+			return 0, 0, err
+		}
+
+		text := strings.TrimSpace(s.Text())
+		if text == "" {
+			continue
+		}
+		parts := strings.Split(text, ":")
+		if len(parts) != 3 {
+			return 0, 0, fmt.Errorf("Cannot parse subuid/gid information: Format not correct for %s file", path)
+		}
+		if parts[0] == username {
```

Comment:
As @LK4D4, we need to return multiple mapping from here.
An example use-case: We have a pre-defined docker-root. We want all other users (except host-root, of course), to be mapped 1:1. 

My /etc/subuid and /etc/subgid looks like this:
```
cat /etc/subuid
docker-root:411460056:3883507239
docker-root:1:411460054
```

Now I start a container: docker run -d ubuntu sleep infinity

Now if I check it's uid_map, 
```
cat /proc/14232/uid_map 
         0  411460056      65536
```